htdocs/accountancy/class/lettering.class.php

```diff
@@ -478,271 +503,353 @@ public function bookkeepingLettering($bookkeeping_ids, $type = 'customer_invoice
 	/**
 	 * Lettering bookkeeping lines
 	 *
-	 * @param	array			$bookkeeping_ids		Lettering specific list of bookkeeping id
-	 * @param	string			$type					Type of bookkeeping type to lettering ('customer_invoice' or 'supplier_invoice')
-	 * @return	array|int								<0 if error otherwise all linked lines by block
+	 * @param	array			$bookkeeping_ids				Lettering specific list of bookkeeping id
+	 * @param	bool			$only_has_subledger_account		Get only lines who have subledger account
+	 * @return	array|int										<0 if error otherwise all linked lines by block
 	 */
-	public function getLinkedLines($bookkeeping_ids, $type = 'customer_invoice')
+	public function getLinkedLines($bookkeeping_ids, $only_has_subledger_account = true)
 	{
 		global $conf, $langs;
 		$this->errors = array();
 
 		// Clean parameters
 		$bookkeeping_ids = is_array($bookkeeping_ids) ? $bookkeeping_ids : array();
-		$type = trim($type);
-
-		if ($type == 'customer_invoice') {
-			$doc_type = 'customer_invoice';
-			$bank_url_type = 'payment';
-			$payment_element = 'paiement_facture';
-			$fk_payment_element = 'fk_paiement';
-			$fk_element = 'fk_facture';
-			$account_number = $conf->global->ACCOUNTING_ACCOUNT_CUSTOMER;
-		} elseif ($type == 'supplier_invoice') {
-			$doc_type = 'supplier_invoice';
-			$bank_url_type = 'payment_supplier';
-			$payment_element = 'paiementfourn_facturefourn';
-			$fk_payment_element = 'fk_paiementfourn';
-			$fk_element = 'fk_facturefourn';
-			$account_number = $conf->global->ACCOUNTING_ACCOUNT_SUPPLIER;
-		} else {
-			$langs->load('errors');
-			$this->errors[] = $langs->trans('ErrorBadParameters');
-			return -1;
-		}
-
-		$payment_ids = array();
 
-		// Get all payment id from bank lines
-		$sql = "SELECT DISTINCT bu.url_id AS payment_id";
+		// Get all bookkeeping lines
+		$sql = "SELECT DISTINCT ab.doc_type, ab.fk_doc";
 		$sql .= " FROM " . MAIN_DB_PREFIX . "accounting_bookkeeping AS ab";
-		$sql .= " LEFT JOIN " . MAIN_DB_PREFIX . "bank_url AS bu ON bu.fk_bank = ab.fk_doc";
-		$sql .= " WHERE ab.doc_type = 'bank'";
-		//	$sql .= " AND ab.subledger_account != ''";
-		//	$sql .= " AND ab.numero_compte = '" . $this->db->escape($account_number) . "'";
-		$sql .= " AND bu.type = '" . $this->db->escape($bank_url_type) . "'";
-		if (!empty($bookkeeping_ids)) $sql .= " AND ab.rowid IN (" . $this->db->sanitize(implode(',', $bookkeeping_ids)) . ")";
-
-		dol_syslog(__METHOD__ . " - Get all payment id from bank lines", LOG_DEBUG);
+		if (!empty($bookkeeping_ids)) {
+			// Get all bookkeeping lines of piece number
+			$sql .= " LEFT JOIN (";
```

Comment:
Because the select does not contains any field of the second table accounting_bookkeeping, the left join on a select distinct is not a good sql.
You should instead use the syntax
AND EXISTS (select rowid from accounting_bookkeeping as pn WHERE ... AND pn.piece_num = ab.piece_num)

Reply:
We have updated the request to work with EXISTS.

Thanks for your return @eldy